Can not create custom procedure with void return type using apoc.custom.declareProcedure

## Guidelines

It is not able to create a custom procedure using  apoc.custom.declareProcedure for void return type . 

dbms.setConfigValue() returns void and custom procedure can not be created . 

```
CALL apoc.custom.declareProcedure(
  'setrole(ssorole::STRING) :: (row::MAP)',
  
  'CALL {CALL dbms.listConfig() YIELD name, value WHERE name CONTAINS "dbms.security.oidc.keycloak.authorization.group_to_role_mapping" RETURN value} 
  WITH value
  CALL dbms.setConfigValue("dbms.security.oidc.keycloak.authorization.group_to_role_mapping",value+$ssorole);'
  
);
```
Throws ==> Failed to invoke procedure `apoc.custom.declareProcedure`: Caused by: java.lang.NullPointerException

## Expected Behavior (Mandatory)

It should create a custom procedure 

## Actual Behavior (Mandatory)

Throws ==> Failed to invoke procedure `apoc.custom.declareProcedure`: Caused by: java.lang.NullPointerException
## How to Reproduce the Problem

Run the above code 

Currently used versions

### Versions

  - OS:
  - Neo4j:4.4.11
  - Neo4j-Apoc:4.4.11

--- a/full/src/main/java/apoc/custom/CypherProcedures.java
+++ b/full/src/main/java/apoc/custom/CypherProcedures.java
@@ -203,14 +203,16 @@ public class CypherProcedures {
     private void checkMode(QueryExecutionType.QueryType queryType, Mode mode) {
         Map<QueryExecutionType.QueryType, Mode> map = Map.of(QueryExecutionType.QueryType.WRITE, Mode.WRITE,
                 QueryExecutionType.QueryType.READ_ONLY, Mode.READ,
-                QueryExecutionType.QueryType.READ_WRITE, Mode.WRITE);
+                QueryExecutionType.QueryType.READ_WRITE, Mode.WRITE,
+                QueryExecutionType.QueryType.DBMS, Mode.DBMS,
+                QueryExecutionType.QueryType.SCHEMA_WRITE, Mode.SCHEMA);
         
         if (!map.get(queryType).equals(mode)) {
             throw new RuntimeException(String.format("The query execution type is %s, but you provided mode %s.\n" +
                             "Supported modes are %s",
                     queryType.name(), 
                     mode.name(), 
-                    map.values().stream().sorted().collect(Collectors.toList()).toString()));
+                    map.values().stream().sorted().collect(Collectors.toList())));
         }
     }
 

--- a/full/src/main/java/apoc/custom/CypherProceduresHandler.java
+++ b/full/src/main/java/apoc/custom/CypherProceduresHandler.java
@@ -287,6 +287,7 @@ public class CypherProceduresHandler extends LifecycleAdapter implements Availab
 
     public static List<FieldSignature> deserializeSignatures(String s) {
         List<Map<String, Object>> mapped = Util.fromJson(s, List.class);
+        if (mapped.isEmpty()) return ProcedureSignature.VOID;
         return mapped.stream().map(map -> {
             String typeString = (String) map.get("type");
             if (typeString.endsWith("?")) {

--- a/full/src/test/java/apoc/custom/CypherProceduresTest.java
+++ b/full/src/test/java/apoc/custom/CypherProceduresTest.java
@@ -3,9 +3,12 @@ package apoc.custom;
 import apoc.RegisterComponentFactory;
 import apoc.SystemLabels;
 import apoc.SystemPropertyKeys;
+import apoc.log.Logging;
+import apoc.result.MapResult;
 import apoc.util.StatusCodeMatcher;
 import apoc.util.TestUtil;
 import org.apache.commons.lang.exception.ExceptionUtils;
+import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -16,6 +19,8 @@ import org.neo4j.graphdb.Node;
 import org.neo4j.graphdb.QueryExecutionException;
 import org.neo4j.graphdb.Transaction;
 import org.neo4j.internal.helpers.collection.Iterators;
+import org.neo4j.procedure.Procedure;
+import org.neo4j.procedure.builtin.BuiltInDbmsProcedures;
 import org.neo4j.test.rule.DbmsRule;
 import org.neo4j.test.rule.ImpermanentDbmsRule;
 
@@ -23,6 +28,8 @@ import java.time.LocalDateTime;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static apoc.custom.CypherProcedures.ERROR_MISMATCHED_INPUTS;
 import static apoc.custom.CypherProcedures.ERROR_MISMATCHED_OUTPUTS;
@@ -49,9 +56,10 @@ public class CypherProceduresTest  {
     @Rule
     public ExpectedException thrown = ExpectedException.none();
 
+
     @Before
     public void setup() {
-        TestUtil.registerProcedure(db, CypherProcedures.class);
+        TestUtil.registerProcedure(db, CypherProcedures.class, BuiltInDbmsProcedures.class);
     }
 
     @Test
@@ -838,6 +846,37 @@ public class CypherProceduresTest  {
         String procedureSignature = "testFail(first::INT, s::INT) :: (answer::INT)";
         assertProcedureFails(String.format(SIGNATURE_SYNTAX_ERROR, procedureSignature), 
                 "call apoc.custom.declareProcedure('" + procedureSignature + "','RETURN $first + $s AS answer')");
+    }
+
+    @Test
+    public void testIssue3349() {
+        String procedure = "CALL apoc.custom.declareProcedure(\n" +
+                "  'retFunctionNames() :: (name :: STRING)',\n" +
+                "  '\n" +
+                "      CALL dbms.functions() YIELD name RETURN name" +
+                "  ',\n" +
+                "  'DBMS'\n" +
+                ");";
+        db.executeTransactionally(procedure);
+        List<String> functions = db.executeTransactionally("CALL custom.retFunctionNames()", Map.of(), result -> result
+                .stream()
+                .map(m -> (String) m.get("name"))
+                .collect(Collectors.toList()));
+        assertFalse(functions.isEmpty());
+
+
+        String procedureVoid = "CALL apoc.custom.declareProcedure(\n" +
+                "  'setTxMetadata(meta :: MAP) :: VOID',\n" +
+                "  '\n" +
+                "      CALL tx.setMetaData($meta)" +
+                "  ',\n" +
+                "  'DBMS'\n" +
+                ");";
+        db.executeTransactionally(procedureVoid);
+        // This should run without exception
+        db.executeTransactionally("CALL custom.setTxMetadata($meta)", Map.of(
+                "meta", Map.of("foo", "bar")
+        ));
     }
     
 

